### PR TITLE
Make torch.svd return V, not V.conj() for complex inputs

### LINKED ISSUE #45821: Singular value decomposition of complex valued matrices does not match mathematical definition
## 📚 Documentation

Now when `torch.svd` supports complex input both for CPU and CUDA. Let's discuss the docs.
Currently, SVD documentation is fully in agreement with the implementation. This function returns`(U, S, V)` such that `input = U @ diag(S) @ V.T`
However, the mathematical definition of SVD is the decomposition of a matrix **M** such that image, where `V^*` is the conjugate transpose of `V`. 

What is the opinion on this? Should the documentation and implementation match math and be updated such that `A = U @ diag(S) @ V.T.conj()`?

[Link to `torch.svd` documentation](https://pytorch.org/docs/stable/generated/torch.svd.html).

cc @ezyang @gchanan @zou3519 @bdhirsh @heitorschueroff @anjali411 @dylanbespalko @mruberry @jlin27 @vishwakftw @jianyuh @nikitaved @pearu @vincentqb

### PR DESCRIPTION
**Pull request by IvanYashchuk**
**BC-breaking note:**

torch.svd() added support for complex inputs in PyTorch 1.7, but was not documented as doing so. The complex "V" tensor returned was actually the complex conjugate of what's expected. This PR fixes the discrepancy.

Note that this means this PR silently breaks all current users of torch.svd() with complex inputs.

**Original PR Summary:**

This PR resolves #45821.

The problem was that when introducing the support of complex inputs for `torch.svd` it was overlooked that LAPACK/MAGMA returns the conjugate transpose of V matrix, not just the transpose of V. So `torch.svd` was silently returning U, S, V.conj() instead of U, S, V.

Behavior of `torch.linalg.pinv`, `torch.pinverse` and `torch.linalg.svd` (they depend on `torch.svd`) is not changed in this PR.

### CODE REVIEW

**Comment by IvanYashchuk** [comment]
[This link points to the current documentation.](https://pytorch.org/docs/master/generated/torch.svd.html?highlight=svd#torch.svd)
And here is the updated rendered documentation. Inconsistent use of italics and code highlight is fixed (in favor of italics), inline math sections are replaced with italics as well.
image
image
image

**Comment by IvanYashchuk** [comment]
XLA's `torch.pinverse` test fails. This test compares the output of normal PyTorch with XLA's PyTorch https://github.com/pytorch/xla/blob/master/test/cpp/test_aten_xla_tensor.cpp#L3341-L3353
XLA's SVD is implemented here https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/client/lib/svd.cc#L808-L878
A possible reason for the failure could be that here I changed `at::pinverse` to use `at::linalg_svd` instead of `at::svd`.

**Comment by mruberry** [comment]
> XLA's `torch.pinverse` test fails. This test compares the output of normal PyTorch with XLA's PyTorch https://github.com/pytorch/xla/blob/master/test/cpp/test_aten_xla_tensor.cpp#L3341-L3353
> XLA's SVD is implemented here https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/client/lib/svd.cc#L808-L878
> A possible reason for the failure could be that here I changed `at::pinverse` to use `at::linalg_svd` instead of `at::svd`.
> A quick fix for this PR would be not to use `linalg_svd` for now until relevant bindings are implemented in pytorch/xla.

cc @ailzhang @JackCaoG 

Continuing to use svd sounds good.

File: test/test_linalg.py [hunk 1]
```diff
@@ -1873,19 +1873,6 @@ def run_subtest(actual_rank, matrix_size, batches, device, svd_lowrank, **option
         actual_rank, size, batches = 2, (17, 4), ()
         run_subtest(actual_rank, size, batches, device, jitted)
 
-    @onlyCPU
-    @skipCPUIfNoLapack
-    @dtypes(torch.cfloat)
-    def test_svd_complex(self, device, dtype):
```
Comment (mruberry) [inline test/test_linalg.py:1879]: Should this test be repurposed to verify the new behavior?

**Comment by mruberry** [comment]
> This PR resolves #45821.
> 
> The problem was that when introducing the support of complex inputs for `torch.svd` it was overlooked that LAPACK/MAGMA returns the conjugate transpose of V matrix, not just the transpose of V. So `torch.svd` was silently returning U, S, V.conj() instead of U, S, V.
> 
> Behavior of `torch.linalg.pinv`, `torch.pinverse` and `torch.linalg.svd` (they depend on `torch.svd`) is not changed in this PR.

Just so I'm confident I understand what's going on here:

- torch.svd has this weird thing where it returns V^T for real-valued inputs, not V
- the generalization of that for complex inputs is that it would return the conjugate transpose of V, V^H
- when complex support was added to torch.svd, it was mistakenly added so that V was returned, not V^H
- this PR address that inconsistency

Follow-up question: how long has torch.svd been returning V and not V^H for complex inputs?

**Comment by IvanYashchuk** [comment]
LAPACK and MAGMA return V.transpose() for real inputs and V.conj().transpose() for complex inputs.
Then in PyTorch output of LAPACK and MAGMA is transposed, so on current master branch `torch.svd` returns V.transpose().transpose() -> V for real inputs and V.conj().transpose().transpose() -> V.conj() for complex inputs.

torch.svd has this weird thing wh

> torch.svd has this weird thing where it returns V^T for real-valued inputs, not V
> the generalization of that for complex inputs is that it would return the conjugate transpose of V, V^H
> when complex support was added to torch.svd, it was mistakenly added so that V was returned, not V^H

* No, torch.svd returns V for real-valued inputs
* The generalization to complex inputs is to return V as well. But it was mistakenly added so that V.conj() is returned, not V
* this PR address that inconsistency

torch.svd had been returning V.conj(), not V since the 3rd of October. Introduced in this PR https://github.com/pytorch/pytorch/pull/45795

File: test/test_linalg.py (hunk 1 above)
Comment (IvanYashchuk) [reply test/test_linalg.py:1879]: Yes, it should. I'll do that.

**Comment by mruberry** [comment]
> LAPACK and MAGMA return V.transpose() for real inputs and V.conj().transpose() for complex inputs.
> Then in PyTorch output of LAPACK and MAGMA is transposed, so on current master branch `torch.svd` returns V.transpose().transpose() -> V for real inputs and V.conj().transpose().transpose() -> V.conj() for complex inputs.
> 
> torch.svd has this weird thing wh
> 
> > torch.svd has this weird thing where it returns V^T for real-valued inputs, not V
> > the generalization of that for complex inputs is that it would return the conjugate transpose of V, V^H
> > when complex support was added to torch.svd, it was mistakenly added so that V was returned, not V^H
> 
> * No, torch.svd returns V for real-valued inputs
> * The generalization to complex inputs is to return V as well. But it was mistakenly added so that V.conj() is returned, not V
> * this PR address that inconsistency
> 
> torch.svd had been returning V.conj(), not V since the 3rd of October. Introduced in this PR #45795

Got it, thank you for clarifying. This all makes sense now. Looks like this support is part of PyTorch 1.7.1, so this will be a BC-breaking change. That's probably fine, I'll just need to run it through some additional tests internally.

**Review by albanD — APPROVED** [review]
Thanks for the cleanup @IvanYashchuk !

The PR looks good to me.

**Pull request closed without merging**